sky/provision/vast/utils.py

```diff
@@ -0,0 +1,103 @@
+# pylint: disable=assignment-from-no-return
+"""Vast library wrapper for SkyPilot."""
+from typing import Any, Dict, List, Optional
+
+from sky import sky_logging
+from sky.adaptors import vast
+
+logger = sky_logging.init_logger(__name__)
+
+
+def list_instances() -> Dict[str, Dict[str, Any]]:
+    """Lists instances associated with API key."""
+    instances = vast.vast().show_instances()
+
+    instance_dict: Dict[str, Dict[str, Any]] = {}
+    for instance in instances:
+        instance['id'] = str(instance['id'])
+        info = instance.copy()
+
+        if isinstance(instance['actual_status'], str):
+            info['status'] = instance['actual_status'].upper()
+        else:
+            info['status'] = 'UNKNOWN'
+        info['name'] = instance['label']
+
+        instance_dict[instance['id']] = info
+
+    return instance_dict
+
+
+def launch(name: str, instance_type: str, region: str, disk_size: int,
+           image_name: str, ports: Optional[List[int]], public_key: str) -> str:
+    """Launches an instance with the given parameters.
+
+    Converts the instance_type to the Vast GPU name, finds the specs for the
+    GPU, and launches the instance.
+    """
+    del ports
+    del public_key
+
+    gpu_name = instance_type.split('-')[1].replace('_', ' ')
+    num_gpus = int(instance_type.split('-')[0].replace('x', ''))
+
+    query = ' '.join([
+        f'geolocation="{region[-2:]}"',
+        f'disk_space>={disk_size}',
+        f'num_gpus={num_gpus}',
+        f'gpu_name="{gpu_name}"',
+    ])
+
+    instance_list = vast.vast().search_offers(query=query)
+
+    if isinstance(instance_list, int) or len(instance_list) == 0:
+        return ''
```

Comment:
Should we raise some error here?

Reply:
Actually I'm a little bit confused when would this triggered. IIUC all instance type should from the catalog, which is not supposed to have 0 offers?

Reply:
This does a real-time inventory search and as such, there's a potentially meaningful cadence delay between the catalog and the spot availability of a particular machine. Our infrastructure is quite dynamic and utilization is pretty high so we need to accommodate for the possibility of not being able to serve a requested machine

Reply:
sure, no problem. Catalog declaration is never availability declaration - I see this problem as endemic and inherent to the nature of cached catalogs (supply from 7 hours ago isn't the current supply) but I'd be happy to add something if you want.

Reply:
see fc69f73d43f4d5edb0207db91962d65aa5ec876f